src/finalizer/finalizer.cpp

```diff
@@ -0,0 +1,208 @@
+				LogStringLine(REPORT_STANDARD, "  Dependent deleted");
+				// Reset the index since we're deleting keys while enumerating
+				dwDepdentsKeyIndex = dwDepdentsKeyIndex > 1 ? dwDepdentsKeyIndex-- : 0;
+
+				// Check if there are any subkeys remaining under the dependents key. If not, we
+				// can uninstall the MSI. We'll recheck the key again in case the MSI fails to clean up the 
+				// provider key to make sure we don't have orphaned keys.
+				hr = RegQueryKey(hkDependentsKey, &cSubKeys, NULL);
+				ExitOnFailure(hr, "Failed to query dependents key.");
+
+				LogStringLine(REPORT_STANDARD, "  Remaining dependents: %i", cSubKeys);
+
+				if (0 == cSubKeys)
+				{
+					// This was the final dependent, so now we can remove the installation if the provider wasn't corrupted and
+					// still contains the product ID.
+					hr = RegReadString(hkProviderKey, NULL, &sczProductId);
+
+					if (E_FILENOTFOUND == hr)
+					{
+						LogStringLine(REPORT_STANDARD, "  No product ID found, provider key: %ls", sczProviderKey);
+						hr = S_OK;
+						break;
+					}
+					else
+					{
+						ExitOnFailure(hr, "Failed to read product ID.");
+					}
+
+					// Let's make sure the product is actually installed. The provider key for an MSI typically
+					// stores the ProductCode, DisplayName, and Version, but by calling into MsiGetProductInfo,
+					// we're doing an implicit detect and getting a property back.
+					hr = WiuGetProductInfo(sczProductId, L"DisplayName", &sczDisplayName);
+					if (SUCCEEDED(hr))
+					{
+						// The provider key *should* have the ProductName and ProductVersion properties, but since
+						// we know it's installed, we just query the installer service. 
+						MsiSetInternalUI(INSTALLUILEVEL_NONE, NULL);
+						hr = WiuConfigureProductEx(sczProductId, INSTALLLEVEL_DEFAULT, INSTALLSTATE_ABSENT, L"MSIFASTINSTALL=7 IGNOREDEPENDENCIES=ALL REBOOT=ReallySuppress", &restart);
+						LogStringLine(REPORT_STANDARD, "  Uninstall of \"%ls\" exited with 0x%.8x", sczDisplayName, hr);
+						ExitOnFailure(hr, "Failed to uninstall MSI package, ProductCode:%ls", sczProductId);
+						
+						// Flag any reboot since we need to return that to the bundle.
+						if (WIU_RESTART_INITIATED == restart || WIU_RESTART_REQUIRED == restart)
+						{
+							*pbRestartRequired = TRUE;
+						}
+					}
+					else if (HRESULT_FROM_WIN32(ERROR_UNKNOWN_PRODUCT) == hr || HRESULT_FROM_WIN32(ERROR_UNKNOWN_PROPERTY) == hr)
+					{
+                        // Possibly a corrupted provider key that wasn't cleaned up. We'll just ignore it.
+						hr = S_OK;
+						LogStringLine(REPORT_STANDARD, "  Product is not installed, ProductCode:%ls", sczProductId);
+					}
+				}
+			}
+		}
+
+		ReleaseRegKey(hkDependentsKey);
+		ReleaseRegKey(hkProviderKey);
+	}
+
+LExit:
+	ReleaseStr(sczDisplayName);
+	ReleaseStr(sczProductId);
+	ReleaseStr(sczProviderKey);
+	ReleaseStr(sczDependentsKey);
+	ReleaseRegKey(hkDependentsKey);
```

Comment:
I don't know regutil (and I can't find it around here), but I would assume that releasing something twice is not nice...
Also - we may call release with null... not sure if that works.

Reply:
regutil is coming from the WiX SDK, defined as ```#define ReleaseRegKey(h) if (h) { ::RegCloseKey(h); h = NULL; }```
strutil (also WiX SDK), is defined as ```#define ReleaseStr(pwz) if (pwz) { StrFree(pwz); }```, though it doesn't make the pointer ```NULL```. 

I was concerned about too many open handles since we could be enumerating thousands of keys/subkeys, but if we hit an error we'd fall into the exit block with an unknown state of the various handles.

Reply:
OK - so it's macro magic... 😢 ... sometime I hate what people do with C++.

Anyway - in that case you're right - it will work.